### DISCUSSION TITLE
brew update -> No developer tools installed.

### DISCUSSION THREAD

**Discussion by lukepuplett** [Homebrew/discussions#2391] [Getting started] [answered]
Hello, I'm new to using a Mac. I setup Homebrew some weeks ago and used it to install something (I can't remember what), so it was working the last time I looked, a little while ago.

I went to install Node.js using `brew install node` and it failed with the message below.

**Note** This weekend I updated to Monterey via the normal OS update thing in macOS.

So I ran `brew update` and I see the same message (printed once).

```
Warning: You are using macOS 12.
We do not provide support for this pre-release version.
You will encounter build failures with some formulae.
Please create pull requests instead of asking for help on Homebrew's GitHub,
Twitter or any other official channels. You are responsible for resolving
any issues you experience while you are running this
pre-release version.

Error: No developer tools installed.
Install the Command Line Tools:
  xcode-select --install

Error: 'git' must be installed and in your PATH!
```

I've never ran `brew update` before, so maybe it never worked and the Monterey thing and the macOS 12 warning are a coincidence. Here's my output from `brew config`.

```
HOMEBREW_VERSION: >=2.5.0 (shallow or no git repository)
ORIGIN: (none)
HEAD: (none)
Last commit: never
Core tap ORIGIN: (none)
Core tap HEAD: (none)
Core tap last commit: never
Core tap branch: (none)
HOMEBREW_PREFIX: /opt/homebrew
HOMEBREW_CASK_OPTS: []
HOMEBREW_MAKE_JOBS: 8
Homebrew Ruby: 2.6.8 => /System/Library/Frameworks/Ruby.framework/Versions/2.6/usr/bin/ruby
CPU: octa-core 64-bit arm_firestorm_icestorm
Clang: N/A
Git: N/A
Curl: 7.77.0 => /usr/bin/curl
macOS: 12.0.1-arm64
CLT: N/A
Xcode: N/A
Rosetta 2: false
```

Is this a problem, or expected and I need to install some that other stuff first?

Thanks

**Comment by carlocab** [comment, marked as answer]
Did you upgrade to macOS 12 recently? Running the Homebrew install script would have installed the Command Line Tools (CLT), but upgrading your OS often removes CLT installations.

Run
```
xcode-select --install
```
and your next `brew update` should work.

**Reply by lukepuplett** [reply]
Thanks for super fast reply. I'll try at home it says it'll take 300h to download and install from this cafe, ha.

**Reply by lukepuplett** [reply]
```
> brew doctor
Your system is ready to brew.
```

Woohoo!

**Reply by yangalli** [reply]
Did it take 300h? Because it's saying the same thing here... 🥺

**Reply by kekkerino** [reply]
What is this doing exactly? Mine says the same.
Edit: it seems that the time actually does go down from 120 hours to 30 minutes.

**Reply by Codegrammer2002** [reply]
I just updated my Mac and it shows 35 hours.

**Reply by Alum-G** [reply]
How long did it take eventually please ? 🙏🏻

**Reply by alextsil** [reply]
about 20min

**Reply by l3v1tat3** [reply]
Mine took about 10 minutes to download & about 5 minutes to install. Although, it started at 30+ hours! 😨

**Reply by jayanthmahadesh** [reply]
thanks

**Reply by Hackbabz69** [reply]
mine is showing 100 hours

**Reply by PannBlackDurian** [reply]
mine is showing 77 hours lol

**Reply by alphashiv27** [reply]
in mine it is showing as -
xcode-select: note: install requested for command line developer tools

**Comment by alifornia** [comment]
Homebrew bottles (binary package) depend on Command Line Tools (CLT) for installation. After an upgrade, CLT should be reinstalled yet it does not mean xcode installation. To check if you have CLT installed? 

`brew config`

If you found `CLT: N\A`. Follow @carlocab's answer. Run: `xcode-select --install`, ...

**Reply by carlocab** [reply]
> To check if you have CLT installed?

`brew doctor` will also tell you if you don't have the CLT installed. After you've installed the CLT, you should also run `brew doctor` anyway to check that your CLT installation was successful. (`xcode-select --install` has been known to occasionally install the wrong version.)

**Comment by erijosilva** [comment]
Here some instructions that solved my problem after upgrade my system:
> $ /bin/bash -c "$(curl -fsSL https://raw.githubusercontent.com/Homebrew/install/HEAD/install.sh)"
> 
> (echo; echo 'eval "$(/opt/homebrew/bin/brew shellenv)"') >> ~/.zprofile
> eval "$(/opt/homebrew/bin/brew shellenv)"
> 
> brew doctor

@see https://mac.install.guide/commandlinetools/3

**Reply by SMillerDev** [reply]
That guide is rubbish. It's pretending Homebrew is a way of installing the commandline tools.
It isn't meant to be, `xcode-select --install` is.

**Reply by erijosilva** [reply]
Only `xcode-select --install` didn't solved it here, bro.

**Reply by rosenjcb** [reply]
> That guide is rubbish. It's pretending Homebrew is a way of installing the commandline tools. It isn't meant to be, `xcode-select --install` is.

Except this was the only thing that fixed my problem after ~30 minutes of Googling. What's funny is that, instead of reading this thread (thus discovering that `xcode-select --install` is a deadend due to various OS changes), you attacked the working solution.

Sounds like you became hyper obsessed over your idea of "what works" and decided to become aggressive in what you think is the solution. This is clown behavior.

**Reply by rosenjcb** [reply]
> That guide is rubbish. It's pretending Homebrew is a way of installing the commandline tools.
> It isn't meant to be, `xcode-select --install` is.

I just can't let this go. It actually baffles me that someone comes here, offers a solution, and instead of engaging with that person and figuring out why they proposed X over Y, you call it rubbish. Homebrew is an open source project, if an outside contributor offers a solution and it's meaningfully helpful to the users, THEN WHAT THE HELL ARE YOU COMPLAINING ABOUT?

But I guess open source is a playground for ego-ridden amateurs that lack basic social skills...

**Reply by p-linnane** [reply]
@rosenjcb Homebrew is not a Command Line Tools installer. A workaround happening to succeed does not make it the correct method.

Your claim that `xcode-select --install` is a dead end is factually incorrect. It is Apple’s own method for installing the Command Line Tools without manually downloading the .pkg. If you can provide a reproducible issue with that path, including your macOS version and exact output, we’ll investigate. Otherwise, stop attacking the volunteers who keep this project alive and move on.

**Reply by rosenjcb** [reply]
> @rosenjcb Homebrew is not a Command Line Tools installer. A workaround happening to succeed does not make it the correct method.

Doesn't make it incorrect either. This is semantics and not meaningful to helping users overcome their local setup issues.

> Your claim that xcode-select --install is a dead end is factually incorrect. It is Apple’s own method for installing the Command Line Tools without manually downloading the .pkg. 

It's obviously a deadend for users in this thread including myself. Maybe there is a working solution to get `xcode-select --install` working, but as several people in this thread have stated, they get a no-op and a small log stating `xcode-select: note: install requested for command line developer tools`.

Only after fixing the broken xcode configuration (via @erijosilva suggestion) and running the xcode-select command again, I can now see a new message: `xcode-select: note: Command line tools are already installed. Use "Software Update" in System Settings or the softwareupdate command line interface to install updates`.

So for several users, probably thousands, they have broken xcode installations (likely from various OS upgrades).

>If you can provide a reproducible issue with that path, including your macOS version and exact output, we’ll investigate. Otherwise, stop attacking the volunteers who keep this project alive and move on.

This is fair. Next time you see a comment, it'll be with ample evidence.

But maybe sit down with @SMillerDev and show him how to meaningfully engage with people when they have problems and offer solutions. Because myself and many others have found benefit/use from @erijosilva input, a lot more than what you or @SMillerDev  have provided so far.

**Reply by p-linnane** [reply]
I’ll make sure to sit down with @SMillerDev and thank him for his years of unpaid work maintaining Homebrew. You do not get to attack our volunteers and then lecture us about how we should serve you. Take a moment to read [Open Source Maintainers Owe You Nothing](https://mikemcquaid.com/open-source-maintainers-owe-you-nothing/).

You were asked to stop and chose to continue. You said, “Next time you see a comment, it’ll be with ample evidence.” There won’t be a next time. You are no longer welcome to participate here and have been banned.
